harvesting/oaipmh.py

```diff
@@ -143,3 +83,76 @@ def get_datacite():
                    encoding='UTF-8',
                    xml_declaration=True,
                    method='xml')
+
+
+def _build_oaipmh_elem():
+    root_elem = Tree.Element('{http://www.openarchives.org/OAI/2.0/}OAI-PMH')
+
+    response_date_elem = Tree.Element('{http://www.openarchives.org/OAI/2.0/}responseDate')
+    response_date_elem.text = datetime.utcnow().replace(microsecond=0).isoformat() + 'Z'
+    root_elem.append(response_date_elem)
+
+    request_elem = Tree.Element('{http://www.openarchives.org/OAI/2.0/}request')
+    request_elem.set('verb', 'ListRecords')
+    request_elem.set('metadataPrefix', 'datacite4')
+    request_elem.text = 'https://zenodo.org/oai2d'
+    root_elem.append(request_elem)
+
+    listrecords_elem = Tree.Element('{http://www.openarchives.org/OAI/2.0/}ListRecords')
+    root_elem.append(listrecords_elem)
+
+    return root_elem
+
+
+def _get_redirect(software):
+    response = requests.head('https://doi.org/{conceptdoi}'.format(conceptdoi=software["conceptDOI"]))
+    if response.status_code == 302:
+        return response.next.url
+    elif response.status_code == 429:
+        print("HttpError 429 while trying to redirect from doi.org.")
+        response.raise_for_status()
+    else:
+        raise HTTPError("Expected a redirect from doi.org to zenodo.org, got {0} instead."
+                        .format(response.status_code))
+
+
+def _get_zenodo_identifier(redirect_url, headers):
+    response = requests.head(redirect_url, headers=headers)
+    if int(response.headers.get('x-ratelimit-remaining', -1)) < 10:
+        # throttle requests
+        logger.info("Sleeping for 60 seconds to avoid HttpError 429")
+        time.sleep(60)
```

Comment:
Code duplication with lines 137 onwards. Suggestion to extract a function here. Name could be _throttle_if_needed but I'm sure a better function name can be found.

Also, isn't 60 seconds a bit long here? How often can this happen during harvesting? If that is many times, harvesting can suddenly take, let's say, hours. Is that correct? If this can be triggered only 1-5 times per harvesting command, this is probably OK.

Reply:
it is currently triggered twice. Zenodo has x-rate-limit 60 per minute for token authenticated users. It's more of a problem for `harvest metadata`, because that returns quicker than `harvest releases`

Also, Erik's branch is about to be merged, he solved some problems already